File: lecs/Lec7_AddingParams/codes/src/tutorial2/TestHelloObject.py
from m5.params import *
from m5.SimObject import SimObject

class TestHelloObject(SimObject):
    type = 'TestHelloObject'
    cxx_header = "tutorial2/testhello_object.hh"

    time_to_wait = Param.Latency("Time before firing the event")
    number_of_fires = Param.Int(1, "Number fo times to fire the event before "
                                   "goodbye")

    testgoodbye_object = Param.TestGoodbyeObject("A testgoodbye object")

class TestGoodbyeObject(SimObject):
    type = 'TestGoodbyeObject'
    cxx_header = "tutorial2/testgoodbye_object.hh"

    buffer_size = Param.MemorySize('1kB',
                                   "Size of buffer to fill with goodbye")
    write_bandwidth = Param.MemoryBandwidth('100MB/s', "Bandwidth to fill "
                                            "the buffer")

File: lecs/Lec7_AddingParams/codes/src/tutorial2/testhello_object.cc
#include "tutorial2/testhello_object.hh"
#include "tutorial2/testgoodbye_object.hh"
#include "debug/TestHello.hh"

TestHelloObject::TestHelloObject(TestHelloObjectParams *params) :
    SimObject(params), event(*this),
    testgoodbye(*(params->testgoodbye_object)),
    myName(params->name),
    latency(params->time_to_wait),
    timesLeft(params->number_of_fires)
{
    DPRINTF(TestHello, "Created the test hello object with the name %s\n", myName);
}


TestHelloObject*
TestHelloObjectParams::create()
{
    return new TestHelloObject(this);
}


void TestHelloObject::processEvent()
{
    timesLeft--;
    DPRINTF(TestHello, "Test Hello World! Processing the event! %d left\n", timesLeft);

    if(timesLeft <= 0) {
        DPRINTF(TestHello, "Done firing!\n");
        testgoodbye.sayGoodbye(myName);
    } else {
        schedule(event, curTick() + latency);
    }
}

void TestHelloObject::startup() 
{
    schedule(event, latency);
}